# Here we will generate a Q&A dataset from a list of corpus sections, to be used later for fine tuning a small llm dedicated to dermatology related questions

### number of non numeric keys (ex 436)

In [43]:
# import re
# strings = [matches[i][0] for i in range(len(matches))]
# def count_non_numeric(strings):
#     return sum(1 for s in strings if not re.fullmatch(r'-?\d+(\.\d+)?(e-?\d+)?', s))

# # Example usage
# non_numeric_count = count_non_numeric(strings)
# print(f"Number of non-numeric strings: {non_numeric_count}")

Number of non-numeric strings: 9


In [1]:
pip install -q -U google-generativeai

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
import google.generativeai as genai

genai.configure(api_key='AIzaSyB4IDge5fr6P89xoTToba8Psdy1jPzUZbk')
model = genai.GenerativeModel('gemini-pro')

### Testing the model

In [3]:
response = model.generate_content("Who is the GOAT of soccer ?")
response.text

'Lionel Messi'

### creating a list of corpus sections  that we can loop over to generate the Q&A dataset 

In [ ]:
import re
import time
import numpy as np
# Define the regex pattern to match key-value pairs
pattern = r'"(.*?)":\s*"(.*?)"'

# Open the text file
with open('llm_dataset.json', 'r') as file:
    # Read the contents of the file
    text = file.read()

    # Find all key-value pairs in the text using regex
    matches = re.findall(pattern, text, re.DOTALL)

    # Loop over the matches and print the values
    lens = []
    for key, value in matches:
        # print(f"{key} ---------")
        lens.append(len(value))
        # print(value.strip())  # Strip any leading or trailing whitespace
        # print()
        # time.sleep(3)
print(len(matches))

# [
#  ("1" : "corpus ...") ,
#  ("2" : "corpus ...") ,
# ....
#  ]


In [8]:
import pickle

with open('sections.pkl', 'rb') as f:
    matches = pickle.load(f)
len(matches)

In [ ]:
with open("gemini_API.txt",'r') as file:
    apis =  [file.readline().strip(),file.readline().strip(),file.readline().strip()]
print(apis)

In [61]:
error_index = [] ## a list that keeps track of corpus sections that could not be queried due to api errors or restrictions
stop_at = 0
with open("qa_dataset.txt","w",encoding='utf-8') as file:
    
    for API_KEY in apis:  
    ## here we try different api (same gemini-pro api with different gmail accounts) 
    ## because of the limited quota with incase the maximum daily api calls is exceeded

        genai.configure(api_key=API_KEY)
        model = genai.GenerativeModel('gemini-pro')
        print(f"using a new api")
        
        for i in range(stop_at,len(matches)) :
            prompt = """given the following text generate 10 questions and their answers from the text, make sure the question varie in complexity, and answers are very detailed, the format must be like this
            {
            "question": .... ,
            "answer": ....
            }
            """ + matches[stop_at][1]
            try:  
                response = model.generate_content(prompt)
                #content = f"\n=========== {stop_at+i}} ===========\n\n"+response.text
                file.write(response.text)
                print(f'section {stop_at} DONE')
                stop_at= stop_at+1
                
            except Exception as e:
                print(f'Exception type: {type(e).__name__}')
                print(f'{type(e).__name__}: {e}')
                if( type(e).__name__ == 'ResourceExhausted' ): # incase the maximum daily api calls is exceeded
                    break
                error_index.append(stop_at)
                stop_at= stop_at+1
                continue


print(error_index)

using a new api
section 0 DONE
section 1 DONE
section 2 DONE
section 3 DONE
section 4 DONE
section 5 DONE
section 6 DONE
section 7 DONE
section 8 DONE
section 9 DONE
section 10 DONE
section 11 DONE
section 12 DONE
section 13 DONE
section 14 DONE
section 15 DONE
section 16 DONE
section 17 DONE
section 18 DONE
section 19 DONE
section 20 DONE
section 21 DONE
section 22 DONE
section 23 DONE
Exception type: ValueError
ValueError: Invalid operation: The `response.text` quick accessor requires the response to contain a valid `Part`, but none were returned. Please check the `candidate.safety_ratings` to determine if the response was blocked.
section 25 DONE
section 26 DONE
section 27 DONE
section 28 DONE
section 29 DONE
section 30 DONE
section 31 DONE
section 32 DONE
section 33 DONE
section 34 DONE
section 35 DONE
section 36 DONE
section 37 DONE
section 38 DONE
section 39 DONE
section 40 DONE
section 41 DONE
section 42 DONE
section 43 DONE
section 44 DONE
section 45 DONE
section 46 DONE
sectio

In [62]:
stop_at

2937

In [65]:
len(error_index)

139

<h4>only 139 sections out of 2937 did not get queried due to gemini api restrictions (certain keywords trigger a safety mechanism)
</h4>


#### more details about gemini blocking mechanism here [here](https://www.googlecloudcommunity.com/gc/AI-ML/Gemini-1-5-Pro-API-blocking-prompt-help/m-p/744895)